v2/bigtable-cdc-to-hbase/src/main/java/com/google/cloud/teleport/v2/templates/transforms/HbaseRowMutationIO.java

```diff
@@ -0,0 +1,224 @@
+      return Objects.hash(configuration, tableId);
+    }
+
+    /**
+     * The writeReplace method allows the developer to provide a replacement object that will be
+     * serialized instead of the original one. We use this to keep the enclosed class immutable. For
+     * more details on the technique see <a
+     * href="https://lingpipe-blog.com/2009/08/10/serializing-immutable-singletons-serialization-proxy/">this
+     * article</a>.
+     */
+    private Object writeReplace() {
+      return new SerializationProxy(this);
+    }
+
+    private static class SerializationProxy implements Serializable {
+      public SerializationProxy() {}
+
+      public SerializationProxy(WriteRowMutations writeRowMutations) {
+        configuration = writeRowMutations.configuration;
+        tableId = writeRowMutations.tableId;
+      }
+
+      private void writeObject(ObjectOutputStream out) throws IOException {
+        SerializableCoder.of(SerializableConfiguration.class)
+            .encode(new SerializableConfiguration(this.configuration), out);
+        StringUtf8Coder.of().encode(this.tableId, out);
+      }
+
+      private void readObject(ObjectInputStream in) throws IOException {
+        this.configuration = SerializableCoder.of(SerializableConfiguration.class).decode(in).get();
+        this.tableId = StringUtf8Coder.of().decode(in);
+      }
+
+      Object readResolve() {
+        return HbaseRowMutationIO.writeRowMutations()
+            .withConfiguration(configuration)
+            .withTableId(tableId);
+      }
+
+      private Configuration configuration;
+      private String tableId;
+    }
+
+    private final Configuration configuration;
+
+    private final String tableId;
+
+    /** Function to write row mutations to an hbase table. */
+    private class WriteRowMutationsFn extends DoFn<KV<byte[], RowMutations>, Void> {
+
+      WriteRowMutationsFn(WriteRowMutations writeRowMutations) {
+        checkNotNull(writeRowMutations.tableId, "tableId");
+        checkNotNull(writeRowMutations.configuration, "configuration");
+      }
+
+      @Setup
+      public void setup() throws Exception {
+        connection = ConnectionFactory.createConnection(configuration);
+        table = connection.getTable(TableName.valueOf(tableId));
+      }
+
+      @StartBundle
+      public void startBundle(StartBundleContext c) throws IOException {
+        recordsWritten = 0;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext c) throws Exception {
+        RowMutations mutations = c.element().getValue();
+
+        // TODO: we use MutateRow(RowMutations). Figure out if it'd be more efficient batch writing here
+        //  with e.g. BufferedMutator.mutate(Mutations) after grouping by rowkey.
+        table.mutateRow(mutations);
+        Metrics.counter("HbaseRepl", "mutations_written_to_hbase").inc();
+        ++recordsWritten;
+      }
+
+      @FinishBundle
+      public void finishBundle() throws Exception {
```

Comment:
We can probably just remove startBundle and finishBundle since we don't do the batching right. This was leftover from the HbaseIO batching in beam

Reply:
removed

Reply:
Used to initialize light-weight connection to Hbase and to log record count for debugging purposes.